source/server/filter_chain_manager_impl.h

```diff
@@ -134,26 +136,54 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
   const Network::Address::InstanceConstSharedPtr address_;
 };
 
-class FilterChainImpl : public Network::FilterChain {
+/**
+ * Base implementation for TCP and QUIC filter chain.
+ */
+class FilterChainImplBase : public Network::FilterChain {
 public:
-  FilterChainImpl(Network::TransportSocketFactoryPtr&& transport_socket_factory,
-                  std::vector<Network::FilterFactoryCb>&& filters_factory)
-      : transport_socket_factory_(std::move(transport_socket_factory)),
-        filters_factory_(std::move(filters_factory)) {}
+  FilterChainImplBase(std::vector<Network::FilterFactoryCb>&& filters_factory)
+      : filters_factory_(std::move(filters_factory)) {}
 
   // Network::FilterChain
-  const Network::TransportSocketFactory& transportSocketFactory() const override {
-    return *transport_socket_factory_;
-  }
-
   const std::vector<Network::FilterFactoryCb>& networkFilterFactories() const override {
     return filters_factory_;
   }
 
 private:
-  const Network::TransportSocketFactoryPtr transport_socket_factory_;
   const std::vector<Network::FilterFactoryCb> filters_factory_;
 };
 
+class TcpFilterChainImpl : public FilterChainImplBase {
+public:
+  TcpFilterChainImpl(Network::TransportSocketFactoryPtr&& transport_socket_factory,
+                     std::vector<Network::FilterFactoryCb>&& filters_factory)
+      : FilterChainImplBase(std::move(filters_factory)),
+        transport_socket_factory_(std::move(transport_socket_factory)) {}
+
+  // Network::FilterChain
+  const Network::TransportSocketFactory* transportSocketFactory() const override {
+    return transport_socket_factory_.get();
+  }
+
+private:
+  const Network::TransportSocketFactoryPtr transport_socket_factory_;
+};
+
+class QuicFilterChainImpl : public FilterChainImplBase {
+public:
+  QuicFilterChainImpl(std::unique_ptr<Ssl::ContextConfig> tls_context_config,
+                      std::vector<Network::FilterFactoryCb>&& filters_factory)
+      : FilterChainImplBase(std::move(filters_factory)),
+        tls_context_config_(std::move(tls_context_config)) {}
+
+  // Network::FilterChain
+  const Network::TransportSocketFactory* transportSocketFactory() const override { return nullptr; }
+
+  Ssl::ContextConfig* tlsContextConfig() const { return tls_context_config_.get(); }
```

Comment:
Where is this one used?

Reply:
not used yet. Will be used to create EnvoyQuicProofSource which is a fake one now.

Reply:
Can this return a const reference or just a reference?

Reply:
Yes